crates/iceberg/src/spec/manifest_list.rs

```diff
@@ -0,0 +1,581 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! ManifestList for Iceberg.
+
+use apache_avro::{from_value, Reader};
+
+use crate::{spec::Literal, Error};
+
+use super::{FormatVersion, StructType};
+
+/// Snapshots are embedded in table metadata, but the list of manifests for a
+/// snapshot are stored in a separate manifest list file.
+///
+/// A new manifest list is written for each attempt to commit a snapshot
+/// because the list of manifests always changes to produce a new snapshot.
+/// When a manifest list is written, the (optimistic) sequence number of the
+/// snapshot is written for all new manifest files tracked by the list.
+///
+/// A manifest list includes summary metadata that can be used to avoid
+/// scanning all of the manifests in a snapshot when planning a table scan.
+/// This includes the number of added, existing, and deleted files, and a
+/// summary of values for each field of the partition spec used to write the
+/// manifest.
+#[derive(Debug, Clone)]
+pub struct ManifestList {
+    version: FormatVersion,
+    /// Entries in a manifest list.
+    entries: Vec<ManifestListEntry>,
+}
+
+impl ManifestList {
+    /// Parse manifest list from bytes.
+    ///
+    /// QUESTION: Will we have more than one manifest list in a single file?
```

Comment:
A question need to discuss. After figure out it, I will remove it.

Reply:
I remembered that @Fokko used to give an answer: YES.

Reply:
Do you mean manifest list entry or manifest list? I think only one manifest list, but several entries? Each entry points to a manifest file?

Reply:
I think there should only be one manifest list with many manifest list entries.

Reply:
This was also confusing for me at first. So we have the manifest list that points to multiple manifests (2 in the example below, one is collapsed):

![image](https://github.com/apache/iceberg-rust/assets/1134248/afc5599c-9ae4-47e5-aee0-df14e9f9fc09)

And then each of the manifests has so-called manifest entries (that point to a Parquet file):

![image](https://github.com/apache/iceberg-rust/assets/1134248/85365f17-957c-4f6d-b172-cb1c7faf809c)

These can be many. A manifest entry in manifest-list is appended each time you do a write operation to the table. For example, if you append to a table, a new manifest is added to the manifest list, and the manifest has references to the newly added files.

Reply:
Thanks for your explaination! I get it!